File: Suyeon/ch10_Deque_Priority_queue/23_Merge_K_Sorted_Lists.py
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:

        heap = []

        for head in lists: # 연결리스트 헤드를 차례대로 받기
            while head:  # 힙에 하나씩 집어넣기 
                heapq.heappush(heap, head.val)
                head = head.next

        result = ListNode() # 결과를 저장할 연결리스트 생성
        p = result

        while heap:  # pop해서 오름차순으로 담기
            p.next = ListNode(heapq.heappop(heap))
            p = p.next

        return result.next

File: Sujin/Ch10_deque_and_priority_queue/23_merge_k_sorted_lists.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

 struct ListNode {
     int val;
     ListNode *next;
     ListNode() : val(0), next(nullptr) {}
     ListNode(int x) : val(x), next(nullptr) {}
     ListNode(int x, ListNode *next) : val(x), next(next) {}
};

struct compare{
    bool operator()(ListNode* &a, ListNode* &b){
        return (a->val) > (b->val); // 오름차순 정렬
    }
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*, vector<ListNode* >, compare> pq;
        
        ListNode* head = new ListNode(0);
        ListNode* tmp = head;
        
        for(int i=0; i<lists.size(); i++){
            if(lists[i]!=NULL) pq.push(lists[i]);
        }
        while(pq.size()>0){
            ListNode* p = pq.top();
            tmp->next = new ListNode(p->val);
            tmp = tmp->next;
            pq.pop();
            if(p->next!=NULL) pq.push(p->next);
        }
        
        return head->next;
    }
};